src/app/reporting/ReportSchedulerImpl.cpp

```diff
@@ -0,0 +1,194 @@
+    }
+    else
+    {
+        Milliseconds32 newTimeout = Milliseconds32(newNode->GetMaxInterval().count() - now.count());
+        // If the handler is not reportable now, schedule a report for the max interval
+        ReturnErrorOnFailure(ScheduleReport(newTimeout, aReadHandler));
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR ReportSchedulerImpl::ScheduleReport(Timeout timeout, ReadHandler * aReadHandler)
+{
+    // Verify the handler is still registered
+    ReadHandlerNode * node = FindReadHandlerNode(aReadHandler);
+    VerifyOrReturnError(nullptr != node, CHIP_ERROR_NOT_FOUND);
+
+    // Cancel Report if it is currently scheduled
+    InteractionModelEngine::GetInstance()->GetExchangeManager()->GetSessionManager()->SystemLayer()->CancelTimer(
+        ReportTimerCallback, node);
+
+    if (!IsChunkedReport(aReadHandler))
+    {
+        // Schedule Report
+        ReturnErrorOnFailure(
+            InteractionModelEngine::GetInstance()->GetExchangeManager()->GetSessionManager()->SystemLayer()->StartTimer(
+                timeout, ReportTimerCallback, node));
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+void ReportSchedulerImpl::CancelReport(ReadHandler * aReadHandler)
+{
+    ReadHandlerNode * node = FindReadHandlerNode(aReadHandler);
+    VerifyOrReturn(nullptr != node);
+    InteractionModelEngine::GetInstance()->GetExchangeManager()->GetSessionManager()->SystemLayer()->CancelTimer(
+        ReportTimerCallback, node);
+}
+
+void ReportSchedulerImpl::UnregisterReadHandler(ReadHandler * aReadHandler)
+{
+    // Verify list is populated and handler is not null
+    VerifyOrReturn((!mReadHandlerList.Empty() || (nullptr == aReadHandler)));
+    CancelReport(aReadHandler);
+
+    ReadHandlerNode * removeNode = FindReadHandlerNode(aReadHandler);
+    // Nothing to remove if the handler is not found in the list
+    VerifyOrReturn(nullptr != removeNode);
+
+    mReadHandlerList.Remove(removeNode);
+    mNodesPool.ReleaseObject(removeNode);
+}
+
+void ReportSchedulerImpl::UnregisterAllHandlers()
+{
+    // Verify list is populated
+    VerifyOrReturn(!mReadHandlerList.Empty());
+    while (!mReadHandlerList.Empty())
+    {
+        ReadHandler * firstReadHandler = mReadHandlerList.begin()->GetReadHandler();
+        UnregisterReadHandler(firstReadHandler);
+    }
+}
+
+bool ReportSchedulerImpl::IsReportScheduled(ReadHandler * aReadHandler)
+{
+    ReadHandlerNode * node = FindReadHandlerNode(aReadHandler);
+    VerifyOrReturnValue(nullptr != node, false);
+    return InteractionModelEngine::GetInstance()->GetExchangeManager()->GetSessionManager()->SystemLayer()->IsTimerActive(
+        ReportTimerCallback, node);
+}
+
+ReportSchedulerImpl::ReadHandlerNode * ReportSchedulerImpl::FindReadHandlerNode(const ReadHandler * aReadHandler)
+{
+    // Verify list is populated and handler is not null
+    VerifyOrReturnValue(!mReadHandlerList.Empty() && (nullptr != aReadHandler), nullptr);
```

Comment:
If none of the read handlers are ever null, or the list is empty, there will be no difference. This particular set of checks can be removed without breaking anything.